### PR TITLE
[select] break the select loop if interrupt_on_signal flag is set

### PR DESCRIPTION
**Pull request by stepanblyschak**
Signed-off-by: Stepan Blyschak <stepanb@nvidia.com>

### CODE REVIEW

**Comment by kcudnik** [comment]
what is actual motivation for this change?

**Comment by stepanblyschak** [comment]
@kcudnik This is to fix https://github.com/Azure/sonic-swss-common/issues/603

**Comment by liat-grozovik** [comment]
@qiluo-msft could you please help to review or suggest a reviewer ?

File: common/configdb.h [hunk 1]
```diff
@@ -88,16 +88,16 @@ class ConfigDBConnector_Native : public SonicV2Connector_Native
...
             # Pass all initial data that we DID NOT send as updates to handlers through the init callback if provided by caller
             if init_data_handler:
                 init_data_handler(init_callback_data)
 
             while not SignalHandlerHelper.checkSignal(SIGNAL_INT):
-                item = self.pubsub.listen_message()
+                item = self.pubsub.listen_message(interrupt_on_signal=True)
                 if 'type' not in item:
-                    # When timeout or cancelled, item will not contains 'type' 
+                    # When timeout or cancelled, item will not contains 'type'
```
Comment (qiluo-msft) [inline common/configdb.h:100]: >cancelled [](http://example.com/codeflow?start=38&length=9)

Please update comment with "interrupt on signal"

File: common/select.cpp [hunk 2]
```diff
@@ -89,26 +89,34 @@ void Select::addSelectables(vector<Selectable *> selectables)
...
-    while(ret == -1 && errno == EINTR && !SignalHandlerHelper::checkSignal(Signals::SIGNAL_INT)); // Retry the select if the process was interrupted by a signal
-
-    if (SignalHandlerHelper::checkSignal(Signals::SIGNAL_INT))
-    {
-        // Return if the epoll_wait was interrupted by SIGTERM
-        return Select::SIGNALINT;
+        // on signal interrupt check if we need to return
+        if (ret == -1 && errno == EINTR)
+        {
+            if (interrupt_on_signal || SignalHandlerHelper::checkSignal(Signals::SIGNAL_INT))
```
Comment (qiluo-msft) [inline common/select.cpp:104]: >SignalHandlerHelper [](http://example.com/codeflow?start=39&length=19)

Can we just remove SignalHandlerHelper class from C++ code?

Comment (stepanblyschak) [reply common/select.cpp]: Yes, removed it

File: common/configdb.h (hunk 1 above)
Comment (stepanblyschak) [reply common/configdb.h]: Done

**Review by qiluo-msft — COMMENTED** [review]
LGTM

**Comment by liuh-80** [comment]
@stepanblyschak , please check my PR for this issue: https://github.com/sonic-net/SONiC/pull/989

My solution is to remove the endless loop in swsscommon API, so any code using swsscommon not need any code change and the issue will be resloved.

File: common/pubsub.cpp [hunk 3]
```diff
@@ -128,13 +128,13 @@ MessageResultPair PubSub::get_message_internal(double timeout)
 
 // Note: it is not straightforward to implement redis-py PubSub.listen() directly in c++
 // due to the `yield` syntax, so we implement this function for blocking listen one message
-std::map<std::string, std::string> PubSub::listen_message()
+std::map<std::string, std::string> PubSub::listen_message(bool interrupt_on_signal)
 {
     const double GET_MESSAGE_INTERVAL = 600.0; // in seconds
     MessageResultPair ret;
     for (;;)
```
[the remark this answers is not in the document]
Comment (qiluo-msft) [reply common/pubsub.cpp:135]: @liuh-80 Is it harm to keep the loop?

File: common/configdb.h [hunk 4]
```diff
@@ -88,16 +88,16 @@ class ConfigDBConnector_Native : public SonicV2Connector_Native
...
+            init_callback_data = {tbl: data for tbl, data in init_data.items() if load_data(tbl, data)}
 
             # Pass all initial data that we DID NOT send as updates to handlers through the init callback if provided by caller
             if init_data_handler:
                 init_data_handler(init_callback_data)
 
-            while not SignalHandlerHelper.checkSignal(SIGNAL_INT):
-                item = self.pubsub.listen_message()
+            while True:
+                item = self.pubsub.listen_message(interrupt_on_signal=True)
```
[the remark this answers is not in the document]
Comment (qiluo-msft) [reply common/configdb.h:98]: I am okay with behavior change on `listen_message`. It is not expected to be used by applications.

**Comment by stepanblyschak** [comment]
> @stepanblyschak , please check my PR for this issue: [sonic-net/SONiC#989](https://github.com/sonic-net/SONiC/pull/989)
> 
> My solution is to remove the endless loop in swsscommon API, so any code using swsscommon not need any code change and the issue will be resloved.

@liuh-80 Which PR are you refering to?

File: common/configdb.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (stepanblyschak) [reply common/configdb.h:98]: @liuh-80 @qiluo-msft PubSub::listen_message is a public API. Anyone, including 3rd party application extensions, may use it. I am not confident changing its default behaviour.

File: common/pubsub.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (stepanblyschak) [reply common/pubsub.cpp:135]: @liuh-80 @qiluo-msft I have the same concerns changing it due to the reason I describe in my previous comment. A behaviour change is a different subject, not relevant to this PR. If there's a decision, that listen_message is not expected to be used by anyone, then this API should be made private and its behaviour may change. That deserves to be a different PR. This PR is a bug fix.

**Comment by liuh-80** [comment]
> 
The most biggest different of my PR is not use the interrupt_on_signal and remove the loop, since keep the loop is necessary, so I will approve this PR:
https://github.com/Azure/sonic-swss-common/pull/629

**Review by liuh-80 — APPROVED** [review]

**Comment by stepanblyschak** [comment]
Now it is the other way around, VS tests are failing but SAIRedis for some reason passed, retrying

**Comment by stepanblyschak** [comment]
These two tests are failing now:

```
test_drop_counters.py::TestDropCounters::test_add_remove_port FAILED     [ 20%]
test_drop_counters.py::TestDropCounters::test_createAndDeleteMultipleCounters FAILED [ 20%]
```

However they pass locally on the same Azure.sonic-swss-common.20220629.6:

```
root@r-sonic-vs-005:~/sonic-swss/tests# py.test test_drop_counters.py -v --imgname=docker-sonic-vs:Azure.sonic-swss-common.20220629.6                                                                                                      ================================================ test session starts =================================================                                                                                                                     platform linux -- Python 3.8.10, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/bin/python3                                                                                                                                                cachedir: .pytest_cache                                                                                                                                                                                                                    rootdir: /root/sonic-swss/tests                                                                                                                                                                                                            plugins: flaky-3.7.0                                                                                                                                                                                                                       collected 17 items                                                                                                                                                                                                                                                                                                                                                                                                                                                                    test_drop_counters.py::TestDropCounters::test_deviceCapabilitiesTablePopulated PASSED                          [  5%]                                                                                                                      test_drop_counters.py::TestDropCounters::test_flexCounterGroupInitialized PASSED                               [ 11%]                                                                                                                      test_drop_counters.py::TestDropCounters::test_createAndRemoveDropCounterBasic PASSED                           [ 17%]                                                                                                                      test_drop_counters.py::TestDropCounters::test_createAndRemoveDropCounterReversed PASSED                        [ 23%]                                                                                                                      test_drop_counters.py::TestDropCounters::test_createCounterWithInvalidCounterType PASSED                       [ 29%]                                                                                                                      test_drop_counters.py::TestDropCounters::test_createCounterWithInvalidDropReason PASSED                        [ 35%]                                                                                                                      test_drop_counters.py::TestDropCounters::test_addReasonToInitializedCounter PASSED                             [ 41%]                                                                                                                      test_drop_counters.py::TestDropCounters::test_removeReasonFromInitializedCounter PASSED                        [ 47%]                                                                                                                      test_drop_counters.py::TestDropCounters::test_removeAllDropReasons PASSED                                      [ 52%]                                                                                                                      test_drop_counters.py::TestDropCounters::test_addDropReasonMultipleTimes PASSED                                [ 58%]
test_drop_counters.py::TestDropCounters::test_addInvalidDropReason PASSED                                      [ 64%]
test_drop_counters.py::TestDropCounters::test_removeDropReasonMultipleTimes PASSED                             [ 70%]
test_drop_counters.py::TestDropCounters::test_removeNonexistentDropReason PASSED                               [ 76%]
test_drop_counters.py::TestDropCounters::test_removeInvalidDropReason PASSED                                   [ 82%]
test_drop_counters.py::TestDropCounters::test_add_remove_port PASSED                                           [ 88%]
test_drop_counters.py::TestDropCounters::test_createAndDeleteMultipleCounters PASSED                           [ 94%]
test_drop_counters.py::test_nonflaky_dummy PASSED                                                              [100%]
```

Looks like they are not stable. Will retry.

**Comment by stepanblyschak** [comment]
Request to cherry-pick to 202205 branch

**Comment by Junchao-Mellanox** [comment]
Hi @qiluo-msft , could you help cherry pick to 202205?

**Comment by qiluo-msft** [comment]
@yxieca will be the right picker for 202205.

**Pull request merged**